libchisel/src/checkexports.rs

```diff
@@ -0,0 +1,177 @@
+use super::ModuleValidator;
+
+use parity_wasm::elements::{
+    ExportEntry, ExportSection, External, FunctionSection, FunctionType, ImportSection, Internal,
+    Module, Type,
+};
+
+/// Enum representing a type of export and any extra data to check.
+pub enum ExportType<'a> {
+    Function(&'a str, FunctionType),
+    Global(&'a str),
+    Memory(&'a str),
+    Table(&'a str),
+}
+
+/// Trait over ExportType that lets a caller check if it is exported in a given module.
+trait isExported {
+    fn is_exported(&self, module: &Module) -> bool;
+}
+
+/// Struct on which ModuleValidator is implemented.
+struct CheckExports<'a> {
+    entries: Vec<ExportType<'a>>,
+    allow_unlisted: bool,
+}
+
+impl<'a> CheckExports<'a> {
+    fn ewasm() -> Self {
+        CheckExports {
+            entries: vec![
+                ExportType::Function("main", FunctionType::default()),
+                ExportType::Memory("memory"),
+            ],
+            allow_unlisted: false,
+        }
+    }
+}
+
+impl<'a> ModuleValidator for CheckExports<'a> {
+    fn validate(self, module: &Module) -> Result<bool, String> {
+        let required_exports_good = !self
+            .entries
+            .iter()
+            .map(|x| x.is_exported(module))
+            .find(|x| *x == false)
+            .is_some();
+        //TODO: check if any extra exports exist
+        Ok(required_exports_good)
+    }
+}
+
+impl<'a> isExported for ExportType<'a> {
+    fn is_exported(&self, module: &Module) -> bool {
+        if let Some(section) = module.export_section() {
+            match self {
+                ExportType::Function(field, sig) => has_func_export(module, field, sig),
+                ExportType::Global(field) => has_global_export(section, field),
+                ExportType::Memory(field) => has_memory_export(section, field),
+                ExportType::Table(field) => has_table_export(section, field),
+            }
+        } else {
+            false
+        }
+    }
+}
+
+//NOTE: has_*_export is implemented with repeating code because you can't implement a trait for
+//enum variants, as they are not types. Furthermore, having one helper for non-func exports would
+//be ugly because information about the export type must still be passed down.
```

Comment:
Nitpicking: here comments don't have a space after the slash. Other comments use three slashes.

Reply:
3 slash comments are for documentation using `rustdoc`. These are just in-source notes. I can make them all use a space, though.